algorithm/pge-eda.metta

```diff
@@ -0,0 +1,564 @@
+                    ($mapped (genPhen $gSpace $cSpace $tSpace $nts $f $maxIter ((py-atom random.random))))
+                    )
+
+                (if (== $t ())
+                    (cons-atom $mapped $t)
+                    (let $r (genPhen $gSpace $cSpace $tSpace $nts $t $maxIter ((py-atom random.random)))
+                        (cons-atom $mapped $r) )))                  
+
+            (let*
+                (   
+                    ($codon (getCodon $tSpace ((py-atom random.random))))
+                    ($nextRule (getRule $start $gSpace $cSpace $codon 0 0 (py-atom random.random)))
+                
+                    (() (println! $nextRule))
+                    (() (updateIndex $tSpace ((py-atom random.random))))
+                    ($codonIndex (match $tSpace (codon-index $i) $i)))
+
+                (if (< $codonIndex $maxIter)
+                    (genPhen $gSpace $cSpace $tSpace $nts $nextRule $maxIter ((py-dot np random.uniform)0 1))
+
+                    X )))))
+
+;; prCounter -- increments a production rule's count by one when called
+;;              $cSpace -- count space
+;;              $nt -- a non-terminal
+;;              $r -- a production rule
+;;              $random -- random number to prevent metta from caching results
+;;                      $c -- count variable
+;;                      $i -- index variable
+;;                      $next -- next index to access the next production rule of a non-terminal
+
+(= (prCounter $cSpace $nt $r $random)
+    (let* 
+        (
+            (($c $i) (match $cSpace ($nt $r $c $i) ($c $i)))
+            (() (remove-atom $cSpace ($nt $r $c $i)))
+            ($next (+ 1 $c)))
+
+        (add-atom $cSpace ($nt $r $next $i))))
+
+;; ;; ::::::::::::: GENERATE GENERATION :::::::::::::
+;; genG -- generate a single generation of size n
+;;              $n -- population size
+;;              $gSpace -- grammar space
+;;              $tSpace -- genotype space
+;;              $genSpace -- space for population
+;;              $maxCount -- number of inidviduals to generate
+;;              $start -- start symbol
+;;              $genId -- genertion id for identification of genrations -- stays the same for a generation of individuals
+;;              $$id -- id for an individual expression -- unique for an expression in a generation
+;;              $random -- needed to prevent metta from caching results of recursive calls              
+;;                      $i -- indvidual expression
+;;                      $g -- genotype tuple used for the expression
+;;                      $c -- production rule count info for an expression collected in a tuple
+
+(= (genG $gSpace $cSpace $tSpace $popSpace $maxCount $start $genId $id $random)
+    (if (< $id $maxCount)
+        (let*
+            (   
+                (() (setGenotype PREC L $tSpace ((py-atom random.random))))
+                ($i (genPhen $gSpace $cSpace $tSpace NTS $start MAXITER ((py-atom random.random))))
```

Comment:
Is the `$i` here individual right? It might get confused with index, can you rename it :)